In [1]:
import pandas as pd
import ast
import numpy as np
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)

In [2]:
df_credits=pd.read_csv('/kaggle/input/the-movies-dataset/credits.csv')
df_keywords=pd.read_csv('/kaggle/input/the-movies-dataset/keywords.csv')
df_movies=pd.read_csv('/kaggle/input/the-movies-dataset/movies_metadata.csv')
df_ratings=pd.read_csv('/kaggle/input/the-movies-dataset/ratings.csv')

In [3]:
df_credits.head(2)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew    id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  8844

In [4]:
df_credits.dtypes

cast    object
crew    object
id       int64
dtype: object

In [5]:
df_credits.isnull().sum()

cast    0
crew    0
id      0
dtype: int64

In [6]:
df_credits['cast'][0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

In [7]:
def convert1(obj):
    count=0
    L=[]
    for i in ast.literal_eval(obj):
        if count!=3:
            L.append(i['name'])
            count+=1
        else :
            break
    return L

In [8]:
df_credits['cast']=df_credits['cast'].apply(convert1)

In [9]:
df_credits.head(2)

cast  \
0             [Tom Hanks, Tim Allen, Don Rickles]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   

                                                crew    id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  8844

In [10]:
df_credits['cast'][0]

['Tom Hanks', 'Tim Allen', 'Don Rickles']

In [11]:
df_credits['crew'][0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

In [12]:
df_credits['crew'][1]

"[{'credit_id': '52fe44bfc3a36847f80a7cd1', 'department': 'Production', 'gender': 2, 'id': 511, 'job': 'Executive Producer', 'name': 'Larry J. Franco', 'profile_path': None}, {'credit_id': '52fe44bfc3a36847f80a7c89', 'department': 'Writing', 'gender': 2, 'id': 876, 'job': 'Screenplay', 'name': 'Jonathan Hensleigh', 'profile_path': '/l1c4UFD3g0HVWj5f0CxXAvMAGiT.jpg'}, {'credit_id': '52fe44bfc3a36847f80a7cdd', 'department': 'Sound', 'gender': 2, 'id': 1729, 'job': 'Original Music Composer', 'name': 'James Horner', 'profile_path': '/oLOtXxXsYk8X4qq0ud4xVypXudi.jpg'}, {'credit_id': '52fe44bfc3a36847f80a7c7d', 'department': 'Directing', 'gender': 2, 'id': 4945, 'job': 'Director', 'name': 'Joe Johnston', 'profile_path': '/fok4jaO62v5IP6hkpaaAcXuw2H.jpg'}, {'credit_id': '52fe44bfc3a36847f80a7cd7', 'department': 'Editing', 'gender': 2, 'id': 4951, 'job': 'Editor', 'name': 'Robert Dalva', 'profile_path': None}, {'credit_id': '573523bec3a368025100062c', 'department': 'Production', 'gender': 0, '

In [13]:
#Extracting director from crew
def convert_dir(obj):
    L=[]
    for i in ast.literal_eval(obj):
        if i['job']=='Director':
            L.append(i['name'])
        else:
            continue
    return L

In [14]:
df_credits['director']=df_credits['crew'].apply(convert_dir)

In [15]:
df_credits.head(3)

cast  \
0             [Tom Hanks, Tim Allen, Don Rickles]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   
2      [Walter Matthau, Jack Lemmon, Ann-Margret]   

                                                crew     id         director  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  [John Lasseter]  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844   [Joe Johnston]  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  [Howard Deutch]

In [16]:
df_credits=df_credits[['id','director','cast']]
df_credits.head(2)

id         director                                            cast
0   862  [John Lasseter]             [Tom Hanks, Tim Allen, Don Rickles]
1  8844   [Joe Johnston]  [Robin Williams, Jonathan Hyde, Kirsten Dunst]

In [17]:
df_credits.isnull().sum()

id          0
director    0
cast        0
dtype: int64

In [18]:
df_keywords.head(3)

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...

In [19]:
df_keywords['keywords'][0]

"[{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]"

In [20]:
def convert(obj):
    L=[]
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [21]:
df_keywords['keywords']=df_keywords['keywords'].apply(convert)

In [22]:
df_keywords.isnull().sum()

id          0
keywords    0
dtype: int64

In [23]:
df_movies.head(2)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

  popularity                       poster_path  \
0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0

In [24]:
df_movies.dtypes

adult                     object
belongs_to_collection     object
budget                    object
genres                    object
homepage                  object
id                        object
imdb_id                   object
original_language         object
original_title            object
overview                  object
popularity                object
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
dtype: object

In [25]:
df_movies.isnull().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

In [26]:
df_movies['revenue'].isnull().sum()

6

In [27]:
df_movies[df_movies['revenue'].isna()]

adult  \
19729                                              False   
19730                                 - Written by Ørnås   
29502                                              False   
29503   Rune Balot goes to a casino connected to the ...   
35586                                              False   
35587   Avalanche Sharks tells the story of a bikini ...   

                                   belongs_to_collection  \
19729                                                NaN   
19730                                           0.065736   
29502  {'id': 122661, 'name': 'Mardock Scramble Colle...   
29503                                           1.931659   
35586                                                NaN   
35587                                           2.185485   

                                 budget  \
19729                                 0   
19730  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29502                                 0   
29503  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35586                                 0   
35587  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19729  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29502  [{'id': 16, 'name': 'Animation'}, {'id': 878, ...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35586  [{'id': 10770, 'name': 'TV Movie'}, {'id': 28,...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage          id  \
19729                                                NaN       82663   
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20   
29502                      http://m-scramble.jp/exhaust/      122662   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29   
35586                                                NaN      249260   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01   

         imdb_id original_language                            original_title  \
19729  tt0113002                en                              Midnight Man   
19730          0             104.0  [{'iso_639_1': 'en', 'name': 'English'}]   
29502  tt2423504                ja                          マルドゥック・スクランブル 排気   
29503          0              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]   
35586  tt2622826                en                          Avalanche Sharks   
35587          0              82.0  [{'iso_639_1': 'en', 'name': 'English'}]   

                                                overview  \
19729  British soldiers force a recently captured IRA...   
19730                                           Released   
29502         Third film of the Mardock Scramble series.   
29503                                           Released   
35586  A group of skiers are terrorized during spring...   
35587                                           Released   

                  popularity                          poster_path  \
19729                    NaN                                  NaN   
19730                    NaN                         Midnight Man   
29502                    NaN                                  NaN   
29503                    NaN  Mardock Scramble: The Third Exhaust   
35586                    NaN                                  NaN   
35587  Beware Of Frost Bites                     Avalanche Sharks   

      production_companies production_countries release_date  revenue  \
19729                  NaN                  NaN          NaN      NaN   
19730                False                  6.0            1      NaN   
29502                  NaN                  NaN          NaN      NaN   
29503                False                  7.0           12      NaN   
35586                  NaN                  NaN          NaN      NaN   
35587                False                  4.3           22      NaN   

   

In [28]:
df_movies.dropna(subset=['revenue'],inplace=True)

In [29]:
df_movies['budget']=df_movies['budget'].astype(float)

In [30]:
#Creating a new column - Movie earnings subtracting budget cost
df_movies['earnings']=df_movies['revenue']-df_movies['budget']

In [31]:
df_movies.head(3)

adult                              belongs_to_collection      budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000.0   
1  False                                                NaN  65000000.0   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0.0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...   
1           Jumanji  When siblings Judy and Peter discover an encha...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...   

  popularity                       poster_path  \
0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2    11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   

              title  video  vote_average  vote_count     earnings  
0         Toy Story  False           7.7      5415.0  343554033.0  
1           Jumanji  False           6.9      2413.0  197797249.0  
2  Grumpier Old Men  False           6.5        92.0          0.0

In [32]:
df_movies.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'earnings'],
      dtype='object')

In [33]:
df_movies.drop(['belongs_to_collection','budget','homepage','imdb_id','poster_path','production_countries','status','tagline','video','original_language','revenue','production_companies','popularity'],axis=1,inplace=True)

In [34]:
df_movies.shape

(45460, 12)

In [35]:
df_movies.isnull().sum()

adult                 0
genres                0
id                    0
original_title        0
overview            954
release_date         84
runtime             257
spoken_languages      0
title                 0
vote_average          0
vote_count            0
earnings              0
dtype: int64

In [36]:
df_movies.dropna(inplace=True)

In [37]:
df_movies.head(3)

adult                                             genres     id  \
0  False  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  False  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  False  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   

     original_title                                           overview  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...   
1           Jumanji  When siblings Judy and Peter discover an encha...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...   

  release_date  runtime                                   spoken_languages  \
0   1995-10-30     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1   1995-12-15    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2   1995-12-22    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   

              title  vote_average  vote_count     earnings  
0         Toy Story           7.7      5415.0  343554033.0  
1           Jumanji           6.9      2413.0  197797249.0  
2  Grumpier Old Men           6.5        92.0          0.0

In [38]:
df_movies['genres']=df_movies['genres'].apply(convert)

In [39]:
df_movies.head(2)

adult                        genres    id original_title  \
0  False   [Animation, Comedy, Family]   862      Toy Story   
1  False  [Adventure, Fantasy, Family]  8844        Jumanji   

                                            overview release_date  runtime  \
0  Led by Woody, Andy's toys live happily in his ...   1995-10-30     81.0   
1  When siblings Judy and Peter discover an encha...   1995-12-15    104.0   

                                    spoken_languages      title  vote_average  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Toy Story           7.7   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...    Jumanji           6.9   

   vote_count     earnings  
0      5415.0  343554033.0  
1      2413.0  197797249.0

In [40]:
df_movies['spoken_languages']=df_movies['spoken_languages'].apply(convert)

In [41]:
df_movies.head(2)

adult                        genres    id original_title  \
0  False   [Animation, Comedy, Family]   862      Toy Story   
1  False  [Adventure, Fantasy, Family]  8844        Jumanji   

                                            overview release_date  runtime  \
0  Led by Woody, Andy's toys live happily in his ...   1995-10-30     81.0   
1  When siblings Judy and Peter discover an encha...   1995-12-15    104.0   

      spoken_languages      title  vote_average  vote_count     earnings  
0            [English]  Toy Story           7.7      5415.0  343554033.0  
1  [English, Français]    Jumanji           6.9      2413.0  197797249.0

In [42]:
df_movies['adult']=df_movies['adult'].astype(bool).astype(int)
df_movies['adult']=df_movies['adult'].replace(1,0)

In [43]:
df_movies.head(2)

adult                        genres    id original_title  \
0      0   [Animation, Comedy, Family]   862      Toy Story   
1      0  [Adventure, Fantasy, Family]  8844        Jumanji   

                                            overview release_date  runtime  \
0  Led by Woody, Andy's toys live happily in his ...   1995-10-30     81.0   
1  When siblings Judy and Peter discover an encha...   1995-12-15    104.0   

      spoken_languages      title  vote_average  vote_count     earnings  
0            [English]  Toy Story           7.7      5415.0  343554033.0  
1  [English, Français]    Jumanji           6.9      2413.0  197797249.0

In [44]:
df_movies['overview']=df_movies['overview'].apply(lambda x:x.split())

In [45]:
df_movies.head(2)

adult                        genres    id original_title  \
0      0   [Animation, Comedy, Family]   862      Toy Story   
1      0  [Adventure, Fantasy, Family]  8844        Jumanji   

                                            overview release_date  runtime  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   1995-10-30     81.0   
1  [When, siblings, Judy, and, Peter, discover, a...   1995-12-15    104.0   

      spoken_languages      title  vote_average  vote_count     earnings  
0            [English]  Toy Story           7.7      5415.0  343554033.0  
1  [English, Français]    Jumanji           6.9      2413.0  197797249.0

In [46]:
df_credits.head(2)

id         director                                            cast
0   862  [John Lasseter]             [Tom Hanks, Tim Allen, Don Rickles]
1  8844   [Joe Johnston]  [Robin Williams, Jonathan Hyde, Kirsten Dunst]

In [47]:
df_keywords.head(2)

id                                           keywords
0   862  [jealousy, toy, boy, friendship, friends, riva...
1  8844  [board game, disappearance, based on children'...

In [48]:
df_movies.head(2)

adult                        genres    id original_title  \
0      0   [Animation, Comedy, Family]   862      Toy Story   
1      0  [Adventure, Fantasy, Family]  8844        Jumanji   

                                            overview release_date  runtime  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   1995-10-30     81.0   
1  [When, siblings, Judy, and, Peter, discover, a...   1995-12-15    104.0   

      spoken_languages      title  vote_average  vote_count     earnings  
0            [English]  Toy Story           7.7      5415.0  343554033.0  
1  [English, Français]    Jumanji           6.9      2413.0  197797249.0

### Merging 

In [49]:
final_df1=pd.merge(df_credits,df_keywords,on='id')

In [50]:
final_df1.head(3)

id         director                                            cast  \
0    862  [John Lasseter]             [Tom Hanks, Tim Allen, Don Rickles]   
1   8844   [Joe Johnston]  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   
2  15602  [Howard Deutch]      [Walter Matthau, Jack Lemmon, Ann-Margret]   

                                            keywords  
0  [jealousy, toy, boy, friendship, friends, riva...  
1  [board game, disappearance, based on children'...  
2  [fishing, best friend, duringcreditsstinger, o...

In [51]:
df_movies.dtypes

adult                 int64
genres               object
id                   object
original_title       object
overview             object
release_date         object
runtime             float64
spoken_languages     object
title                object
vote_average        float64
vote_count          float64
earnings            float64
dtype: object

In [52]:
df_movies['id']=df_movies['id'].astype(int)
df_movies['release_date']=pd.to_datetime(df_movies['release_date'])

In [53]:
df_movies.head(1)

adult                       genres   id original_title  \
0      0  [Animation, Comedy, Family]  862      Toy Story   

                                            overview release_date  runtime  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   1995-10-30     81.0   

  spoken_languages      title  vote_average  vote_count     earnings  
0        [English]  Toy Story           7.7      5415.0  343554033.0

In [54]:
final_df=pd.merge(final_df1,df_movies,on='id')

In [55]:
final_df.head(2)

id         director                                            cast  \
0   862  [John Lasseter]             [Tom Hanks, Tim Allen, Don Rickles]   
1  8844   [Joe Johnston]  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   

                                            keywords  adult  \
0  [jealousy, toy, boy, friendship, friends, riva...      0   
1  [board game, disappearance, based on children'...      0   

                         genres original_title  \
0   [Animation, Comedy, Family]      Toy Story   
1  [Adventure, Fantasy, Family]        Jumanji   

                                            overview release_date  runtime  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   1995-10-30     81.0   
1  [When, siblings, Judy, and, Peter, discover, a...   1995-12-15    104.0   

      spoken_languages      title  vote_average  vote_count     earnings  
0            [English]  Toy Story           7.7      5415.0  343554033.0  
1  [English, Français]    Jumanji           6.9      2413.0  197797249.0

In [56]:
final_df.columns

Index(['id', 'director', 'cast', 'keywords', 'adult', 'genres',
       'original_title', 'overview', 'release_date', 'runtime',
       'spoken_languages', 'title', 'vote_average', 'vote_count', 'earnings'],
      dtype='object')

In [57]:
df1=final_df[['id','title','overview','genres','keywords','cast','director']]

In [58]:
df1.head(1)

id      title                                           overview  \
0  862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   

                        genres  \
0  [Animation, Comedy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   

                                  cast         director  
0  [Tom Hanks, Tim Allen, Don Rickles]  [John Lasseter]

In [59]:
df1['genres']=df1['genres'].apply(lambda x:[i.replace(" ",'')for i in x])
df1['keywords']=df1['keywords'].apply(lambda x:[i.replace(" ",'')for i in x])
df1['cast']=df1['cast'].apply(lambda x:[i.replace(" ",'')for i in x])
df1['director']=df1['director'].apply(lambda x:[i.replace(" ",'')for i in x])

In [60]:
df1.head(2)

id      title                                           overview  \
0   862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  8844    Jumanji  [When, siblings, Judy, and, Peter, discover, a...   

                         genres  \
0   [Animation, Comedy, Family]   
1  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [boardgame, disappearance, basedonchildren'sbo...   

                                          cast        director  
0             [TomHanks, TimAllen, DonRickles]  [JohnLasseter]  
1  [RobinWilliams, JonathanHyde, KirstenDunst]   [JoeJohnston]

### Creating a new column concattenating all columns except id and title

In [61]:
df1['strings']=df1['overview']+df1['genres']+df1['keywords']+df1['cast']+df1['director']

In [62]:
df1.head(1)

id      title                                           overview  \
0  862  Toy Story  [Led, by, Woody,, Andy's, toys, live, happily,...   

                        genres  \
0  [Animation, Comedy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   

                               cast        director  \
0  [TomHanks, TimAllen, DonRickles]  [JohnLasseter]   

                                             strings  
0  [Led, by, Woody,, Andy's, toys, live, happily,...

In [63]:
new_df=df1[['id','title','strings']]
new_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                             strings  
0  [Led, by, Woody,, Andy's, toys, live, happily,...  
1  [When, siblings, Judy, and, Peter, discover, a...  
2  [A, family, wedding, reignites, the, ancient, ...  
3  [Cheated, on,, mistreated, and, stepped, on,, ...  
4  [Just, when, George, Banks, has, recovered, fr...

### Converting the list data type to string

In [64]:
new_df['strings']=new_df['strings'].apply(lambda x:" ".join(x))

In [65]:
new_df['strings'][2]

"A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the locals who worry she'll scare the fish away. But she's less interested in seafood than she is in cooking up a hot time with Max. Romance Comedy fishing bestfriend duringcreditsstinger oldmen WalterMatthau JackLemmon Ann-Margret HowardDeutch"

### Converting the text in strings column to lower case

In [66]:
new_df['strings']=new_df['strings'].apply(lambda x:x.lower())

In [67]:
new_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                             strings  
0  led by woody, andy's toys live happily in his ...  
1  when siblings judy and peter discover an encha...  
2  a family wedding reignites the ancient feud be...  
3  cheated on, mistreated and stepped on, the wom...  
4  just when george banks has recovered from his ...

## Text vectorisation

### Converting the text into vectors which will be needed for calculation cosine similarity
### Removing the stop words like and,the,are,of etc which won't be required for our project

In [68]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=5000,stop_words='english')

In [69]:
vectors=cv.fit_transform(new_df['strings']).toarray()

In [70]:
cv.fit_transform(new_df['strings']).toarray().shape

(45558, 5000)

In [71]:
cv.get_feature_names()

['000',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '17th',
 '18',
 '18th',
 '19',
 '1920s',
 '1930',
 '1930s',
 '1940',
 '1940s',
 '1941',
 '1942',
 '1943',
 '1944',
 '1945',
 '1950',
 '1950s',
 '1960',
 '1960s',
 '1964',
 '1968',
 '1969',
 '1970',
 '1970s',
 '1971',
 '1973',
 '1974',
 '1976',
 '1979',
 '1980',
 '1980s',
 '1989',
 '1990',
 '1990s',
 '1994',
 '1999',
 '19th',
 '19thcentury',
 '20',
 '200',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '20th',
 '21',
 '21st',
 '22',
 '23',
 '24',
 '25',
 '28',
 '30',
 '300',
 '35',
 '3d',
 '40',
 '50',
 '500',
 '60',
 '60s',
 '70',
 '70s',
 '80',
 '80s',
 '90',
 'aaron',
 'abandoned',
 'abandons',
 'abducted',
 'abilities',
 'ability',
 'able',
 'aboard',
 'abortion',
 'abroad',
 'absence',
 'absurd',
 'abuse',
 'abused',
 'abusive',
 'academy',
 'accept',
 'acceptance',
 'accepted',
 'accepts',
 'access',
 'accident',
 'accidental',
 'accidentally',
 'a

### These are same type of words like began,begins,begin etc. We will apply stammen
### Stammen will replace same type of words into root word eg.'beat','beaten','beating';The root word would be beat

In [72]:
import nltk
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

In [73]:
def stem(text):
    y=[]
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [74]:
new_df['strings']=new_df['strings'].apply(stem)

In [75]:
new_df['strings'][31]

"in the year 2035, convict jame cole reluctantli volunt to be sent back in time to discov the origin of a deadli viru that wipe out nearli all of the earth' popul and forc the survivor into underground communities. but when cole is mistakenli sent to 1990 instead of 1996, he' arrest and lock up in a mental hospital. there he meet psychiatrist dr. kathryn railly, and patient jeffrey goines, the son of a famou viru expert, who may hold the key to the mysteri rogu group, the armi of the 12 monkeys, thought to be respons for unleash the killer disease. sciencefict thriller mysteri schizophrenia philadelphia cassandrasyndrom stockholmsyndrom timetravel post-apocalypt dystopia lethalviru monkey subterranean brucewilli madeleinestow bradpitt terrygilliam"

### We have 45558 movies i.e.., 45558 vectors and in every vector we have 5000 numbers. We need to calculate the distance between each vector with every other vector. We can do this using Cosine Distance, it calculated the distance between vectors by calculating the angle between the vectors.
### Cosine distance and cosine similarity are inversely proportional i.e.., more the distance,less is the similarity between vectors;here movies

In [76]:
from sklearn.metrics.pairwise import cosine_similarity

In [77]:
similarity=cosine_similarity(vectors)

In [78]:
similarity

array([[1.        , 0.048795  , 0.05423261, ..., 0.        , 0.01767767,
        0.        ],
       [0.048795  , 1.        , 0.07938842, ..., 0.        , 0.03450328,
        0.04279605],
       [0.05423261, 0.07938842, 1.        , ..., 0.        , 0.01917412,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.01767767, 0.03450328, 0.01917412, ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.04279605, 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [79]:
sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[1:6]

[(15446, 0.40069384267237695),
 (3004, 0.3879701448994927),
 (10354, 0.3529432893783948),
 (8381, 0.3408228586736446),
 (43632, 0.32791291789197646)]

In [80]:
def recommend(movie):
    movie_index=new_df[new_df['title']==movie].index[0]
    distance=similarity[movie_index]
    movies_list=sorted(list(enumerate(distance)),reverse=True,key=lambda x:x[1])[1:6]
    for i in movies_list:
        print(new_df.iloc[i[0]].title)

In [81]:
recommend('Avatar')

The War in Space
The War in Space
Aliens vs Predator: Requiem
Abraxas, Guardian of the Universe
Project Moon Base


In [82]:
recommend('Batman Begins')

Batman: The Dark Knight Returns, Part 1
The Dark Knight
Batman: Under the Red Hood
Bablo
Batman & Robin


### Credits : Sonali Thakur
https://www.linkedin.com/feed/update/urn:li:activity:7017390515587313664/